databricks/koalas/namespace.py

```diff
@@ -1859,35 +1870,46 @@ def resolve_func(kdf, this_column_labels, that_column_labels):
                     )
                 )
 
-    column_labelses_of_kdfs = [kdf._internal.column_labels for kdf in objs]
+    column_labels_of_kdfs = [kdf._internal.column_labels for kdf in objs]
     if ignore_index:
         index_names_of_kdfs = [[] for _ in objs]
     else:
         index_names_of_kdfs = [kdf._internal.index_names for kdf in objs]
     if all(name == index_names_of_kdfs[0] for name in index_names_of_kdfs) and all(
-        idx == column_labelses_of_kdfs[0] for idx in column_labelses_of_kdfs
+        idx == column_labels_of_kdfs[0] for idx in column_labels_of_kdfs
     ):
         # If all columns are in the same order and values, use it.
         kdfs = objs
     else:
         if join == "inner":
-            interested_columns = set.intersection(*map(set, column_labelses_of_kdfs))
+            interested_columns = set.intersection(*map(set, column_labels_of_kdfs))
             # Keep the column order with its firsts DataFrame.
-            merged_columns = sorted(
-                list(
-                    map(
-                        lambda c: column_labelses_of_kdfs[0][column_labelses_of_kdfs[0].index(c)],
-                        interested_columns,
-                    )
-                )
-            )
+            merged_columns = [
+                label for label in column_labels_of_kdfs[0] if label in interested_columns
+            ]
+
+            # When multi-index column, although pandas is flaky if `join="inner" and sort=False`,
+            # always sort to follow the `join="outer"` case behavior.
+            if (len(merged_columns) > 0 and len(merged_columns[0]) > 1) or sort:
+                merged_columns = sorted(merged_columns)
 
             kdfs = [kdf[merged_columns] for kdf in objs]
         elif join == "outer":
             # If there are columns unmatched, just sort the column names.
-            merged_columns = sorted(
-                list(set(itertools.chain.from_iterable(column_labelses_of_kdfs)))
-            )
+            merged_columns = column_labels_of_kdfs[0].copy()
```

Comment:
no big deal since it's already guarded above but we could add an `assert`  for the length of `column_labels_of_kdfs` to be non-empty.

Reply:
We could also add an assert on `merged_columns` too.